[ID-1363] Add transaction id to resolveDrs requests

samanehsan opened the pull request:
Ticket: https://broadworkbench.atlassian.net/browse/ID-1363

For every request to the resolveDrs endpoint, drshub sends a request to the drs provider (such as TDR) to get object info and optionally another request to get the access url. We need a way to join the Drshub and TDR requests in Bard so that we can then gather drs metrics using the billing profile/dataset/snapshot info.

Changes:
- Copied the `UserLoggingMetrics` class from TDR to allow additional Bard properties to be specified outside of the `TrackingInterceptor` class
- Log a transaction id (a random UUID) to Bard for resolveDrs requests
- Set this transaction id as a header on the requests to TDR (when getting the object metadata and access url)

samanehsan committed:
Add a transaction id header

The transaction id header will be used
to connect a drs resolution request in drshub
to the requests that it makes to TDR, by logging
both in the Bard event properties.

samanehsan committed:
Log transaction id to Bard

samanehsan committed:
Fix pact test

tlangs commented on service/src/main/java/bio/terra/drshub/services/DrsResolutionService.java:
```diff
@@ -235,6 +249,7 @@ DrsObject fetchObjectInfo(
     log.info(drsRequestLogMessage);
 
     var drsApi = drsApiFactory.getApiFromUriComponents(uriComponents, drsProvider);
+    drsApi.setHeader("x-transaction-id", transactionId);
```
`X-Transaction-ID` could probably be extracted into a variable since its referenced multiple times.

tlangs replied:
I don't know if we want to add the transaction id header to every request. Other DRS Providers might have similar header restrictions to us, and not allow it through. We should probably only add it if we know the request is going to TDR.

I think I might have said the opposite when we talked about it face to face, but I'm realizing now that if we're restricting what headers are allowed in requests to our systems, others are probably doing it too.

samanehsan replied:
We're not checking the provider for other headers btw 😅 At least we always pass the IP address and x-user-project for access url requests.

tlangs replied:
Oh, maybe it's fine then!

tlangs commented on service/src/main/java/bio/terra/drshub/services/DrsResolutionService.java:
```diff
@@ -44,17 +46,21 @@ public class DrsResolutionService {
     this.drsProviderService = drsProviderService;
     this.authService = authService;
     this.auditLogger = auditLogger;
+    this.userLoggingMetrics = userLoggingMetrics;
```
DRSHub already logs its DRS Resolutions via the Audit Logger. I think we should include the transaction ID in that message, rather than publishing a separate message.

samanehsan replied:
This allows us to include the transaction id in the properties for existing Bard events (like how TDR does it). It's not creating an additional log message. I can include it in the audit log message as well though!

tlangs replied:
Oh, do we need the `UserLoggingMetrics` so that we can inject more attributes into the Bard request?

samanehsan replied:
Yes! Maybe there's a clearer name for this?

tlangs commented on service/src/main/java/bio/terra/drshub/services/DrsResolutionService.java:
```diff
@@ -266,7 +281,8 @@ AccessURL fetchDrsObjectAccessUrl(
+      String googleProject,
+      String transactionId) {
 
     var drsApi = drsApiFactory.getApiFromUriComponents(uriComponents, drsProvider);
```
You could probably support setting the header conditionally if its TDR from within `getApiFromUriComponents`

tlangs commented:
Also, its convention to Capital-Kebab-Case header names. Libraries usually handle them in a case-insensitive way.

samanehsan committed:
Use Capital-Kebab-Case for header name

samanehsan committed:
Fix logging transactionId

The transaction id needs to be
added to the bard properties in
the same thread that makes the
call to bard (which has the request
context)

samanehsan committed:
Fix tests

samanehsan committed:
Fix GCS API controller test

samanehsan committed:
Fix more tests
